### PR TITLE
support unary-minus on IntegerRangeType

### PR DESCRIPTION
**Pull request by staabm**
refs https://github.com/phpstan/phpstan/issues/5614

### CODE REVIEW

File: tests/PHPStan/Analyser/data/integer-range-types.php [hunk 1]
```diff
@@ -278,4 +278,19 @@ public function math($i, $j, $z, $pi, $r1, $r2, $rMin, $rMax, $x, $y) {
...
+	 * @param int<-5, 10> $r2
+	 * @param int<min, 5> $rMin
+	 * @param int<5, max> $rMax
+	 */
+	public function unaryMinus($r1, $r2, $rMin, $rMax) {
+
+		assertType('int<-10, -1>', -$r1);
+		assertType('int<-10, 5>', -$r2);
+		assertType('int<-5, max>', -$rMin);
+		assertType('int<min, -5>', -$rMax);
```
Comment (orklah) [inline tests/PHPStan/Analyser/data/integer-range-types.php:293]: Would a range with a `0` bound end up with `-0` ?

Comment (staabm) [reply tests/PHPStan/Analyser/data/integer-range-types.php:294]: yep, added a test-case.

**Comment by staabm** [comment]
failling test is unrelated. just rebased.

File: src/Analyser/MutatingScope.php [hunk 2]
```diff
@@ -1019,6 +1020,15 @@ private function resolveType(Expr $node): Type
 				return TypeCombinator::union(...$newTypes);
 			}
 
+			if ($type instanceof IntegerRangeType) {
+				$negativeRange = $this->resolveType(new Node\Expr\BinaryOp\Mul($node->expr, new LNumber(-1)));
+
+				if ( $negativeRange instanceof IntegerRangeType && ($negativeRange->getMin() === null || $negativeRange->getMax() === null)) {
```
Comment (staabm) [inline src/Analyser/MutatingScope.php:1026]: Actually I was not sure whether multiplication by -1 should have already returned the range as we need it and this if-case actually compensates a bug in the Mul logic?

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php:1026]: I don't know, you tell me :) Feel free to fix more bugs here.

Comment (staabm) [reply src/Analyser/MutatingScope.php:1026]: Is `-$b === $b * -1`?

Comment (orklah) [reply src/Analyser/MutatingScope.php:1026]: Seems like it. Here's a small script I made to generate equations for [Wolfram](https://www.wolframalpha.com/input/?i=%28interval%5B-%E2%88%9E%2C2%5D+*+-1%29+%3D+-%28interval%5B-%E2%88%9E%2C2%5D%29%3B): https://3v4l.org/AUHH7

**Comment by staabm** [comment]
@ondrejmirtes we can merge here as is. The multiplication bug will be fixed separately

**Comment by staabm** [comment]
@ondrejmirtes we can merge here as is. The multiplication bug will be fixed separately

**Comment by ondrejmirtes** [comment]
Thank you!

File: src/Analyser/MutatingScope.php (hunk 2 above)
Comment (staabm) [reply src/Analyser/MutatingScope.php:1026]: Fix in https://github.com/phpstan/phpstan-src/pull/670

Comment (staabm) [reply src/Analyser/MutatingScope.php:1026]: btw: the same is true for division: https://www.wolframalpha.com/input/?i=%28interval%5B-%E2%88%9E%2C%E2%88%9E%5D+%2F+-2%29+%3D+-%28interval%5B-%E2%88%9E%2C%E2%88%9E%5D%29%3B

**Pull request merged**